django/db/backends/oracle/operations.py

```diff
@@ -258,9 +258,16 @@ def last_executed_query(self, cursor, sql, params):
         # https://cx-oracle.readthedocs.io/en/latest/cursor.html#Cursor.statement
         # The DB API definition does not define this attribute.
         statement = cursor.statement
-        # Unlike Psycopg's `query` and MySQLdb`'s `_executed`, CxOracle's
-        # `statement` doesn't contain the query parameters. refs #20010.
-        return super().last_executed_query(cursor, statement, params)
+        # Unlike Psycopg's `query` and MySQLdb`'s `_executed`, cx-Oracle's
+        # `statement` doesn't contain the query parameters. Substitute
+        # parameters manually.
+        if isinstance(params, (tuple, list)):
+            for i, param in enumerate(params):
+                statement = statement.replace(':arg%d' % i, force_str(param, errors='replace'))
+        elif hasattr(params, 'keys'):
```

Comment:
It feels a bit odd to _duck-type_ on `.keys` but use `.items` in the loop; maybe `hasattr(params, 'items')` should be used instead? Is there a reason not to use `isinstance(params, dict)` instead?

Reply:
:thinking: Probably not, I change this :+1: